### ISSUE TITLE
Migration to Neutralino.js

### ISSUE THREAD

**Issue by CosmoMyzrailGorynych** [ct-js/ct-js#538]
**Describe the current state of the problem**
Nw.js is based on a seemingly good idea of merging node.js and webkit contexts, but, in hindsight, it is a terrible idea as new APIs come and nw.js starts breaking everywhere with conflicting types, non-working methods, hacks to make this or that libraries work, and just crashing and hanging up just because.

Issues with nw.js block me from updating to pixi.js v8, and updating nw.js breaks pixi.js completely. There are also issues with multiple windows and communication between them, and probably a couple others I've already forgotten.

**Describe the solution you'd like**
Ct.js is already migrating to Neutralino.js. This issue is more for letting others what's going on and for documenting my progress.

**Describe alternatives you've considered (optional)**
Socketsupply.co seemed like a nice library but it is too early to use.
Tauri is a time-proven solution but introducing Rust will make the stack more complicated to develop upon.

**Comment by CosmoMyzrailGorynych** [comment]
### App architecture

Obviously some functions would be lost if using only Neutralino; it has functions for working with filesystem and or other native functions, but other features are lacking or missing, and are critical:

* multi-window communication;
* hosting a static web server (for testing out games);
* npm packages that require node.js API (archiving, file conversion, CORS fetching, etc.)

To handle that, I decided to pair Neutralino.js with Bun.

* Bun is cool and hip and I believe that this time it will be a more serious JS runtime than Deno.
* More performant than Deno or Node.js.
* Has built-in funtionality to cross-build executables of your app, which will simplify building ct.js.
* Has opportunities to upgrade the build pipeline of ct.js so everything runs on one JS runtime. In other words, maybe we can build ct.js with Bun only, without node.js.

With bun, contrary to nw.js, ct.js will become a multi-process application. Extensions like Bun chat with Neutralino through WebSockets, and Neutralino implements its functions in the same way by sending commands to its main process with websockets and webview library.
The implementation for the Bun extension started from https://github.com/hschneider/neutralino-ext-bun and was heavily modified to make promise-based API on both sides.

Neutralino doesn't support cross-window messaging out of the box. Cross-window communication was required due to inability to arbitrarily hook devtools into a random iframe for the game debugger, and plus I wanted to make a separate floating debug toolbar from the very beginning. (But it crashed nw.js while transferring nothing but strings and JSONs.) A novel solution was needed, which is called `multiwindow` and I will probably publish it as a Neutralino extension later. Basically it allows sharing Neutralino's websocket connections between windows and thus manipulating those windows connection tokens you know.

#### IPC map

Each Neutralino node is a window's name in `multiwindow`, and labels on arrows are events they send to the target window.
```mermaid
flowchart LR

ide[<code>ide</code>\nNeutralino window]
game[<code>game</code>\nNeutralino window]
debugToolbar[<code>debugToolbar</code>\nNeutralino window]
qrCodes[<code>qrCodes</code>\nNeutralino window]
bun[Bun subprocess\n]

bun -.desktopBuildProgress.->ide
bun <-- bunchat --> ide

ide <-- stopDebugging --> game & debugToolbar
ide -- reloadGame --> game
debugToolbar -- openDebugExternal --> ide
debugToolbar -- debugActions + e.detail --> game
game -- gameEvents + e.detail --> debugToolbar
debugToolbar -- close --> qrCodes
```

The `qrCodes` window is not yet implemented (a modal for the old "mobile" button in debug tools in nw.js), and maybe a new window will spawn from `debugToolbar` to switch rooms. (Can't create system context menus in Neutralino, and it has to render outside of the toolbar's boundaries, so the solution is to make a modal window with a room list.)

### Progress

Current progress as seen in my private notebook:

![изображение](https://github.com/user-attachments/assets/d7e17bf2-24fc-431e-ac04-7557c00f21ef)

TL;DR: It's good! I've implemented all the questionable features, prototypes worked, and now I'm implementing the meat of the engine back. Projects open, can be edited, and are exported. Polyfills for filesystem and for local file loading were added. Operations not possible with Neutralino are handled by Bun.

**Comment by CosmoMyzrailGorynych** [comment]
Quick status update:
 
* I have to turn the current structure inside out (One main Bun process that manages Neutralino processes), as the current implementation is flaky
* Most stuff was already in this or that form, I just need to recombine the pieces in a different way
